{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "8f526e4b-49cc-4c5f-9b40-ca455b36b7d1",
   "metadata": {},
   "source": [
    "# SlideSeq Spatial Transcriptomics analysis of Muscle Regeneration \n",
    "\n",
    "As described at:\n",
    "Integrated single-cell and spatial transcriptomic analysis of skeletal muscle regeneration \n",
    "Lauren Walter , Jessica Orton , E. H. Hannah Fong , Ms Viviana Maymi , Dr Brian Rudd , Dr Jennifer Elisseeff , Ioannis Ntekas , Iwijn De Vlaminck, Benjamin D. Cosgrove* \n",
    "\n",
    "\n",
    "ST analysis by Ioannis Ntekas \n",
    "\n",
    "email: in68@cornell.edu"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3fbd585a-4afb-4ef1-be24-caa5bf7bbd24",
   "metadata": {},
   "source": [
    "## Import Packages\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "86fbbcba-9c4b-4297-b5c8-6708830d2f0c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import squidpy as sq\n",
    "import numpy as np\n",
    "import anndata as ad\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import os\n",
    "import seaborn as sns\n",
    "import matplotlib as mpl\n",
    "from matplotlib import rcParams\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib.ticker import FuncFormatter\n",
    "from mpl_toolkits.axes_grid1.anchored_artists import AnchoredSizeBar\n",
    "import matplotlib.font_manager as fm\n",
    "import matplotlib.patches as mpatches\n",
    "from sklearn.neighbors import NearestNeighbors\n",
    "from scipy.sparse import csr_matrix\n",
    "import scipy.spatial as scisp\n",
    "from matplotlib.colors import ListedColormap\n",
    "from scipy.stats import ttest_ind, mannwhitneyu, ranksums\n",
    "\n",
    "\n",
    "\n",
    "import gc\n",
    "\n",
    "\n",
    "###import functions from ST_utils.py\n",
    "import ST_utils\n",
    "\n",
    "gc.enable()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ba738fd0-b5af-415c-9c20-0f528aa21bce",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.settings.verbosity = 3  \n",
    "# verbosity: errors (0), warnings (1), info (2), hints (3)\n",
    "sc.logging.print_versions()\n",
    "sc.settings.set_figure_params(dpi=80, facecolor='white', frameon=True, figsize=(5, 5),vector_friendly=True, dpi_save=300)\n",
    "# Font properties for the scale bar\n",
    "fontprops = fm.FontProperties(size=10)\n",
    "# Set global figure parameters for scanpy\n",
    "sc.settings.set_figure_params(fontsize=7)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "21ef4cea-5463-427d-9fa0-53d1c514e134",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define colormap to use with light grey as the zero/ND value \n",
    "viridis = mpl.colormaps['viridis_r'].resampled(256)\n",
    "newcolors = viridis(np.linspace(0, 1, 256))\n",
    "light_grey = np.array([192/256, 192/256, 192/256, 1])\n",
    "newcolors[:1, :] = light_grey\n",
    "newcmp = ListedColormap(newcolors)\n",
    "\n",
    "\n",
    "# Define colormap to use with light grey as the zero/ND value \n",
    "Reds = mpl.colormaps['Reds'].resampled(256)\n",
    "newcolors = Reds(np.linspace(0, 1, 256))\n",
    "light_grey = np.array([192/256, 192/256, 192/256, 1])\n",
    "newcolors[:1, :] = light_grey\n",
    "newcmp2 = ListedColormap(newcolors)\n",
    "\n",
    "# Define colormap to use with light grey as the zero/ND value \n",
    "Magma = mpl.colormaps['magma'].resampled(256)\n",
    "newcolors = Magma(np.linspace(0, 1, 256))\n",
    "light_grey = np.array([192/256, 192/256, 192/256, 1])\n",
    "newcolors[:1, :] = light_grey\n",
    "newcmp3 = ListedColormap(newcolors)\n",
    "\n",
    "\n",
    "# Define colormap to use with light grey as the zero/ND value \n",
    "inferno = mpl.colormaps['inferno'].resampled(256)\n",
    "newcolors = inferno(np.linspace(0, 1, 256))\n",
    "light_grey = np.array([192/256, 192/256, 192/256, 1])\n",
    "newcolors[:1, :] = light_grey\n",
    "newcmp4 = ListedColormap(newcolors)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "266e4571-abd2-4360-aaf2-9c96da6576ee",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the Young and Geriatric object on a dictionary for easier handling 0:Y, 1:G\n",
    "\n",
    "adata_dict = {}\n",
    "Sample_list = [11,12]\n",
    "\n",
    "for i in[0,1]:\n",
    "    temp = sc.read_10x_mtx(f\"/fs/cbsuvlaminck2/workdir/in68/Muscle/out/Sample{Sample_list[i]}/STARsolo/Solo.out/GeneFull/raw/\")\n",
    "    temp.write_h5ad(f\"/fs/cbsuvlaminck2/workdir/in68/Muscle/out/Sample{Sample_list[i]}/STARsolo/Solo.out/GeneFull/raw/Sample_{Sample_list[i]}.h5ad\")\n",
    "    temp.raw = temp.copy()\n",
    "    sc.pp.filter_cells(temp, min_counts = 20)\n",
    "    # add cell2location prediction scores and related vars to the object --copy the .obsm, .uns \n",
    "    adata =sc.read_h5ad (f\"/workdir/in68/Muscle/cell2location/cell2location_out/Muscle_{Sample_list[i]}_0301.h5ad\")\n",
    "    temp.obsm = adata.obsm\n",
    "    temp.uns = adata.uns\n",
    "    adata_dict[i] = temp\n",
    "    adata_dict[i].var_names_make_unique()\n",
    "\n",
    "    \n",
    "    adata_dict[i].obs[adata_dict[i].uns['mod']['factor_names']] = adata_dict[i].obsm['q05_cell_abundance_w_sf']\n",
    "    adata_dict[i].obsm['q05_cell_proportions'] = adata_dict[i].obsm['q05_cell_abundance_w_sf'].div(adata_dict[i].obsm['q05_cell_abundance_w_sf'].sum(axis=1), axis=0)\n",
    "\n",
    "    adata_dict[i].obs[\"total_abundance\"] = adata_dict[i].obsm['q05_cell_abundance_w_sf'].sum(axis = 1)\n",
    "    # adata_dict[i].obs[adata_dict[i].uns['mod']['factor_names']] = adata_dict[i].obsm['q05_cell_abundance_w_sf']\n",
    "    adata_dict[i].obs[adata_dict[i].uns['mod']['factor_names']] = adata_dict[i].obsm['q05_cell_proportions']\n",
    "    adata_dict[i].obs[\"max_pred\"] = adata_dict[i].obs[adata_dict[i].uns['mod']['factor_names']].max(axis=1)\n",
    "    #Set max_pred_celltype as the dominant celltype based on cell2loc results\n",
    "    adata_dict[i].obs[\"max_pred_celltype\"] = adata_dict[i].obs[adata_dict[i].uns['mod']['factor_names']].idxmax(axis=1)\n",
    "    adata_dict[i].obs[\"max_pred_celltype\"] = adata_dict[i].obs[\"max_pred_celltype\"].astype('category')\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "    #####Add spatial coordinates and then filer cells with less that 10 neighbors at 100um\n",
    "    # Add spatial coords\n",
    "    adata_dict[i] = add_spatial_coordinates(adata_dict[i], f\"/fs/cbsuvlaminck2/workdir/in68/Utils/Curio_BB/A0018_0{Sample_list[i]}_BeadBarcodes.txt\")\n",
    "    # Convert distances to um\n",
    "    adata_dict[i].obsm['spatial'] = adata_dict[i].obsm['spatial'] /1.54\n",
    "    # Calculate euclidean distances\n",
    "    adata_dict[i].obsm[\"spatial_distances\"] = scisp.distance.squareform(scisp.distance.pdist(adata_dict[i].obsm['spatial']))\n",
    "    # Filter smear/singlets\n",
    "    adata_dict[i].obs[\"spatial_neighbors_100_true\"] = np.sum(adata_dict[i].obsm[\"spatial_distances\"] < 100, axis=0)\n",
    "    adata_dict[i] = adata_dict[i][adata_dict[i].obs[\"spatial_neighbors_100_true\"] > 10]\n",
    "    \n",
    "    # QC and filtering\n",
    "    sc.pp.filter_genes(adata_dict[i], min_cells = 10)\n",
    "\n",
    "    adata_dict[i].var[\"mito\"] = adata_dict[i].var_names.str.startswith(\"Mt-\")\n",
    "    sc.pp.calculate_qc_metrics(adata_dict[i], qc_vars=[\"mito\"], inplace=True)\n",
    "\n",
    "    # Normalization\n",
    "    sc.pp.normalize_total(\n",
    "            adata_dict[i], \n",
    "            target_sum=1e4,\n",
    "        inplace=True)\n",
    "    sc.pp.log1p(adata_dict[i])\n",
    "    \n",
    "del temp \n",
    " "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1ee34838-3245-4a3b-ad48-993fc1652045",
   "metadata": {},
   "source": [
    "### Define injury loci\n",
    "Injury locus deffinition based on immune cell density and location"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e29d297f-c29d-4ba3-8877-22785ca8c4ed",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define injury locus based on the kNN approach. Calculate immune neighbors \n",
    "\n",
    "for adata in adata_dict.values():\n",
    "\n",
    "    for celltype in np.unique(adata.obs['max_pred_celltype']):\n",
    "        adata = calculate_celltype_neighbors(adata, 300, celltype, 'max_pred_celltype')\n",
    "    \n",
    "    adata.obs['Immune_neighbors'] = adata.obs['T cells_neighbors'] + adata.obs['B cells_neighbors'] + adata.obs['NK cells_neighbors'] + adata.obs['Dendritic cells_neighbors'] + adata.obs['Neutrophils_neighbors'] + adata.obs['Monocytes/Macrophages_neighbors']\n",
    "gc.collect()   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7ae9e91c-eca1-41f7-8134-30d5ab4efb97",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Geriatric Muscle\n",
    "\n",
    "mask = (adata_dict[1].obs[\"Immune_neighbors\"] > 20) & (adata_dict[1].obsm[\"spatial\"][:,0] > 1945)\n",
    "adata_dict[1].obs['injury_classification'] = ''\n",
    "idx = adata_dict[1][mask].obs.index\n",
    "idx_rev = adata_dict[1][~mask].obs.index\n",
    "adata_dict[1].obs.loc[idx_rev, 'injury_classification'] = 'outside'\n",
    "adata_dict[1].obs.loc[idx, 'injury_classification'] = 'injury'\n",
    "\n",
    "sq.pl.spatial_scatter(adata_dict[1],color =\"injury_classification\", size= 1, shape=None,figsize= [5,5],cmap=newcmp,vmin = 0.4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4a4b2342-fece-4663-924d-a368dde60b78",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Young Muscle \n",
    "\n",
    "mask = (adata_dict[0].obs[\"Immune_neighbors\"] > 20) & (adata_dict[0].obsm[\"spatial\"][:,1] < 1500)\n",
    "adata_dict[0].obs['injury_classification'] = ''\n",
    "idx = adata_dict[0][mask].obs.index\n",
    "idx_rev = adata_dict[0][~mask].obs.index\n",
    "adata_dict[0].obs.loc[idx_rev, 'injury_classification'] = 'outside'\n",
    "adata_dict[0].obs.loc[idx, 'injury_classification'] = 'injury'\n",
    "\n",
    "sq.pl.spatial_scatter(adata_dict[0],color =\"injury_classification\", size= 0.1, shape=None,figsize= [5,5],cmap=newcmp,vmin = 0.4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0bd217b0-af9f-40d0-b186-75031af199d6",
   "metadata": {},
   "outputs": [],
   "source": [
    "##Discard singlets/small islands\n",
    "for adata in adata_dict.values():\n",
    "    adata= calculate_celltype_neighbors(adata, 20, celltype_label='injury',cell_obs='injury_classification')\n",
    "    mask = adata.obs[\"injury_neighbors\"] > 10\n",
    "    idx = adata[mask].obs.index\n",
    "    idx_rev = adata[~mask].obs.index\n",
    "    adata.obs.loc[idx_rev, 'injury_classification'] = 'outside'\n",
    "    adata.obs.loc[idx, 'injury_classification'] = 'injury'\n",
    "    \n",
    "sq.pl.spatial_scatter(adata_dict[0],color =\"injury_classification\", size= 1, shape=None,figsize= [5,5],cmap=newcmp,vmin = 0.4)\n",
    "sq.pl.spatial_scatter(adata_dict[1],color =\"injury_classification\", size= 1, shape=None,figsize= [5,5],cmap=newcmp,vmin = 0.4)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cddd12e0-5c5a-41c3-ad5d-477ae02d27eb",
   "metadata": {},
   "source": [
    "### Define cell type and injury locus colors "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4fee2078-7eca-4b7c-882f-7cebb9973162",
   "metadata": {},
   "outputs": [],
   "source": [
    "celltype_colors = [\n",
    "    '#F20A53',\n",
    "    '#9C6DA5',\n",
    "    '#2E3B65',\n",
    "    '#3C0AAA',\n",
    "    '#8BA690',\n",
    "    '#067F71',\n",
    "    '#891B5A',\n",
    "    '#FA78FA',\n",
    "    '#C85A00',\n",
    "    '#076594',\n",
    "    '#709EB5',\n",
    "    '#826E00',\n",
    "    '#5078FA',\n",
    "    '#AFE4DE',\n",
    "    '#793EA8'\n",
    "]  \n",
    "loc_colors =  ['#EC4F38','#2B8076']\n",
    "\n",
    "\n",
    "\n",
    "\n",
    " \n",
    "adata_dict[0].uns['max_pred_celltype_colors'] = celltype_colors\n",
    "adata_dict[0].uns['loc_colors'] = loc_colors\n",
    "adata_dict[1].uns['max_pred_celltype_colors'] = celltype_colors\n",
    "adata_dict[1].uns['loc_colors'] = loc_colors"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8c60cdb7-88f2-4632-8aa8-9bf55842e849",
   "metadata": {},
   "source": [
    "### Make barplots for cell abundances in spatial data "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ff86c467-bdfb-4c07-90ee-1a8365eabb7e",
   "metadata": {},
   "outputs": [],
   "source": [
    "###Store normalization info as uns\n",
    "for i in[0,1]:\n",
    "    total_beads = len(adata_dict[i].obs)\n",
    "    injury = len(adata_dict[i][adata_dict[i].obs['injury_classification'] == 'injury' ].obs)\n",
    "    outside = len(adata_dict[i][adata_dict[i].obs['injury_classification'] != 'injury' ].obs)\n",
    "    \n",
    "    adata_dict[i].uns['total_beads'] = total_beads\n",
    "    adata_dict[i].uns['injury'] = injury\n",
    "    adata_dict[i].uns['outside'] = outside\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "802ad696-c6b1-4127-b173-b752a3b805bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "os.chdir('/workdir/in68/Muscle/figs/')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "849f8796-d0d0-49b5-a9f4-458c9712ad15",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Custom font settings\n",
    "plt.rcParams.update({'font.size': 7})  # Set global font size\n",
    "\n",
    "# Function to format the y-axis ticks as percentages\n",
    "def to_percentage(x, pos):\n",
    "    return f'{int(x)}%'\n",
    "\n",
    "formatter = FuncFormatter(to_percentage)\n",
    "\n",
    "# Define the subset of cell types to include in the plot in the specified order\n",
    "cell_types_subset = ['MuSCs', 'Fusing Myocytes', 'Monocytes/Macrophages', 'B cells', 'T cells', 'Dendritic cells', 'FAPs', 'Endothelial cells']\n",
    "\n",
    "# Create a figure with subplots\n",
    "fig, axes = plt.subplots(nrows=1, ncols=2, figsize=(6.2, 2.5), dpi=300)  # Total size of the figure\n",
    "fig.subplots_adjust(left=0.1, right=0.9, top=0.9, bottom=0.1, wspace=0.15)  # Adjust space to optimize layout\n",
    "\n",
    "for i, location in enumerate(np.unique(adata_dict[0].obs['injury_classification'])):\n",
    "    category = location\n",
    "    colors = ['#F08205', '#4C026E']\n",
    "    dataset_names = ['Young', 'Geriatric']\n",
    "    \n",
    "    cell_type_counts = {}\n",
    "    for idx, (key, adata) in enumerate(adata_dict.items()):\n",
    "        adata = adata[adata.obs['injury_classification'] == category].copy()\n",
    "        counts = adata.obs['max_pred_celltype'].value_counts().rename(dataset_names[idx]).filter(items=cell_types_subset)\n",
    "        cell_type_counts[dataset_names[idx]] = counts / adata.uns[category] * 100\n",
    "    \n",
    "    # Convert the counts to a DataFrame for easier plotting\n",
    "    df_counts = pd.DataFrame(cell_type_counts).fillna(0).reindex(cell_types_subset).reset_index()\n",
    "    df_counts = df_counts.rename(columns={'index': 'max_pred_celltype'}).melt(id_vars='max_pred_celltype', var_name='Dataset', value_name='Count')\n",
    "    \n",
    "    # Plotting\n",
    "    ax = axes[i]\n",
    "    sns.barplot(data=df_counts, x='max_pred_celltype', y='Count', hue='Dataset', palette=colors, edgecolor=None, dodge=True, ax=ax)\n",
    "    ax.set_xticklabels(ax.get_xticklabels(), rotation=90, ha='right')\n",
    "    ax.set_ylim(0, 20)  # Setting a consistent y-axis limit\n",
    "    ax.yaxis.set_major_formatter(formatter)  # Set y-axis labels to show percentages\n",
    "\n",
    "    # Explicit font size settings for elements\n",
    "    ax.tick_params(axis='both', which='major', labelsize=7)\n",
    "    ax.set_xlabel(ax.get_xlabel(), fontsize=7)\n",
    "    ax.set_ylabel(ax.get_ylabel(), fontsize=7)\n",
    "    ax.legend_.remove()  # Remove the legend\n",
    "    ax.set_xlabel('')  # Remove x-axis label\n",
    "\n",
    "    if i == 1:\n",
    "        ax.set_ylabel('')  # Remove y-axis label for the second plot\n",
    "        ax.set_yticklabels([])  # Remove y-axis tick labels for the second plot\n",
    "        \n",
    "\n",
    "# Adjust layout and save\n",
    "plt.tight_layout()  # Auto-adjust layout\n",
    "#plt.savefig('Cell_Type_Comparison_1.pdf', bbox_inches='tight', dpi=300)  # Save with tight bounding box\n",
    "plt.show()  # Show the plot\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cd023c70-1166-4e59-9bb0-69284e9ef3ac",
   "metadata": {},
   "outputs": [],
   "source": [
    "formatter = FuncFormatter(to_percentage)\n",
    "# Define the subset of cell types to include in the plot in the specified order\n",
    "cell_types_subset = ['Myonuclei']\n",
    "\n",
    "# Create a figure with subplots\n",
    "fig, axes = plt.subplots(nrows=1, ncols=2, figsize=(1.5, 2.5), dpi=300)  # Narrower figure as less content needs to be displayed\n",
    "fig.subplots_adjust(left=0.1, right=0.9, top=0.9, bottom=0.1, wspace=0.4)  # Adjust space to optimize layout\n",
    "\n",
    "for i, location in enumerate(np.unique(adata_dict[0].obs['injury_classification'])):\n",
    "    category = location\n",
    "    colors = ['#F08205', '#4C026E']\n",
    "    dataset_names = ['Young', 'Geriatric']\n",
    "    \n",
    "    cell_type_counts = {}\n",
    "    for idx, (key, adata) in enumerate(adata_dict.items()):\n",
    "        adata = adata[adata.obs['injury_classification'] == category].copy()\n",
    "        counts = adata.obs['max_pred_celltype'].value_counts().rename(dataset_names[idx]).filter(items=cell_types_subset)\n",
    "        cell_type_counts[dataset_names[idx]] = counts / adata.uns[category] * 100\n",
    "    \n",
    "    # Convert the counts to a DataFrame for easier plotting\n",
    "    df_counts = pd.DataFrame(cell_type_counts).fillna(0).reindex(cell_types_subset).reset_index()\n",
    "    df_counts = df_counts.rename(columns={'index': 'max_pred_celltype'}).melt(id_vars='max_pred_celltype', var_name='Dataset', value_name='Count')\n",
    "    \n",
    "    # Plotting\n",
    "    ax = axes[i]\n",
    "    sns.barplot(data=df_counts, x='max_pred_celltype', y='Count', hue='Dataset', palette=colors, edgecolor=None, dodge=True, ax=ax)\n",
    "    ax.set_xticklabels(['Myonuclei'], rotation=90, ha='right')  # Use long placeholder to ensure space is similar\n",
    "    ax.set_ylim(0, 100)  # Setting a consistent y-axis limit\n",
    "    ax.yaxis.set_major_formatter(formatter)  # Set y-axis labels to show percentages\n",
    "    \n",
    "    ax.legend_.remove()  # Remove the legend\n",
    "    \n",
    "    ax.tick_params(axis='both', which='major', labelsize=7)\n",
    "    ax.set_xlabel('')  # Remove x-axis label\n",
    "    ax.set_ylabel(ax.get_ylabel(), fontsize=7)\n",
    "    \n",
    "    if i == 1:\n",
    "        ax.set_ylabel('')  # Remove y-axis label for the second plot\n",
    "        ax.set_yticklabels([])  # Remove y-axis tick labels for the second plot\n",
    "\n",
    "# Adjust layout and save\n",
    "plt.tight_layout()  # Auto-adjust layout\n",
    "#plt.savefig('Cell_Type_Comparison_2.pdf', bbox_inches='tight', dpi=300)  # Save with tight bounding box\n",
    "plt.show()  # Show the plot\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bf317883-e47b-4ded-b05c-9eb42d6e6a3d",
   "metadata": {},
   "source": [
    "### Calculate senescence score "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b81ddbac-52c9-4a7b-b2df-e9c607bdf38f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load senescence score genes \n",
    "senescence = pd.read_csv('/workdir/in68/Muscle/metadata/FBR_Gene_List.csv')\n",
    "\n",
    "gene_list = senescence[senescence['logFC'] > 0]['Gene (mouse)'].astype('str')\n",
    "print(gene_list)\n",
    "\n",
    "\n",
    "\n",
    "### Filter gene list \n",
    "muscs = {}\n",
    "i = 0\n",
    "for adata in adata_dict.values():\n",
    "    muscs[i] = adata[adata.obs['max_pred_celltype'] == 'MuSCs']\n",
    "    i = i+1\n",
    "sc.pp.filter_genes(muscs[0],min_cells= 2)\n",
    "sc.pp.filter_genes(muscs[1],min_cells= 2)\n",
    "\n",
    "common_genes = np.intersect1d(muscs[1].var_names.astype('str'),muscs[0].var_names.astype('str'))\n",
    "\n",
    "score_genes = np.intersect1d(common_genes,gene_list)\n",
    "\n",
    "\n",
    "sc.tl.score_genes(adata_dict[0],gene_list= score_genes, score_name= 'Senescence_score',use_raw=False,n_bins= 1000,ctrl_size=50)\n",
    "sc.tl.score_genes(adata_dict[1],gene_list= score_genes, score_name= 'Senescence_score',use_raw=False, n_bins= 1000, ctrl_size=50)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "578cf990-50de-4bc3-b516-24b27bb88167",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define groups for comparison\n",
    "groups = {\n",
    "    'Young_m': (adata_dict[0], adata_dict[0].obs['max_pred_celltype'] == 'MuSCs', 'Young', 'MuSCs_all'),\n",
    "    'Geriatric_m': (adata_dict[1], adata_dict[1].obs['max_pred_celltype'] == 'MuSCs', 'Geriatric', 'MuSCs_all'),\n",
    "    'Young_m_injury': (adata_dict[0], (adata_dict[0].obs['max_pred_celltype'] == 'MuSCs') & (adata_dict[0].obs['injury_classification'] == 'injury'), 'Young', 'MuSCs_injury'),\n",
    "    'Geriatric_m_injury': (adata_dict[1], (adata_dict[1].obs['max_pred_celltype'] == 'MuSCs') & (adata_dict[1].obs['injury_classification'] == 'injury'), 'Geriatric', 'MuSCs_injury'),\n",
    "    'Young_m_noninjury': (adata_dict[0], (adata_dict[0].obs['max_pred_celltype'] == 'MuSCs') & (adata_dict[0].obs['injury_classification'] != 'injury'), 'Young', 'MuSCs_noninjury'),\n",
    "    'Geriatric_m_noninjury': (adata_dict[1], (adata_dict[1].obs['max_pred_celltype'] == 'MuSCs') & (adata_dict[1].obs['injury_classification'] != 'injury'), 'Geriatric', 'MuSCs_noninjury'),\n",
    "    'Young_non': (adata_dict[0], adata_dict[0].obs['max_pred_celltype'] != 'MuSCs', 'Young', 'non_MuSCs_all'),\n",
    "    'Geriatric_non': (adata_dict[1], adata_dict[1].obs['max_pred_celltype'] != 'MuSCs', 'Geriatric', 'non_MuSCs_all'),\n",
    "    'Young_non_injury': (adata_dict[0], (adata_dict[0].obs['max_pred_celltype'] != 'MuSCs') & (adata_dict[0].obs['injury_classification'] == 'injury'), 'Young', 'non_MuSCs_injury'),\n",
    "    'Geriatric_non_injury': (adata_dict[1], (adata_dict[1].obs['max_pred_celltype'] != 'MuSCs') & (adata_dict[1].obs['injury_classification'] == 'injury'), 'Geriatric', 'non_MuSCs_injury'),\n",
    "    'Young_non_noninjury': (adata_dict[0], (adata_dict[0].obs['max_pred_celltype'] != 'MuSCs') & (adata_dict[0].obs['injury_classification'] != 'injury'), 'Young', 'non_MuSCs_noninjury'),\n",
    "    'Geriatric_non_noninjury': (adata_dict[1], (adata_dict[1].obs['max_pred_celltype'] != 'MuSCs') & (adata_dict[1].obs['injury_classification'] != 'injury'), 'Geriatric', 'non_MuSCs_noninjury')\n",
    "}\n",
    "\n",
    "# Extract data and compare each Young vs Geriatric group\n",
    "final_data = []\n",
    "comparisons = [\n",
    "    ('Young_m', 'Geriatric_m'),\n",
    "    ('Young_m_injury', 'Geriatric_m_injury'),\n",
    "    ('Young_m_noninjury', 'Geriatric_m_noninjury'),\n",
    "    ('Young_non', 'Geriatric_non'),\n",
    "    ('Young_non_injury', 'Geriatric_non_injury'),\n",
    "    ('Young_non_noninjury', 'Geriatric_non_noninjury')\n",
    "]\n",
    "\n",
    "# Loop through each comparison, extract data, plot, and calculate statistics\n",
    "for group1, group2 in comparisons:\n",
    "    df1 = extract_data(*groups[group1])\n",
    "    df2 = extract_data(*groups[group2])\n",
    "    final_data.append(df1)\n",
    "    final_data.append(df2)\n",
    "    \n",
    "    print(f\"\\nComparing {group1} vs {group2}\")\n",
    "    # print(f\"{group1}: Mean = {np.mean(df1['Senescence_score'])}, Median = {np.median(df1['Senescence_score'])}, N = {len(df1)}\")\n",
    "    # print(f\"{group2}: Mean = {np.mean(df2['Senescence_score'])}, Median = {np.median(df2['Senescence_score'])}, N = {len(df2)}\")\n",
    "    \n",
    "    # plot_boxplot_for_arrays(df1['Senescence_score'], df2['Senescence_score'], array_1=group1, array_2=group2)\n",
    "    \n",
    "    results = compare_datasets(df1['Senescence_score'], df2['Senescence_score'], n_comparisons=1)\n",
    "    for test, value in results.items():\n",
    "        print(f\"{test}: {value}\")\n",
    "\n",
    "# Combine all data into a single DataFrame\n",
    "final_df = pd.concat(final_data)\n",
    "\n",
    "# Reorder columns\n",
    "final_df = final_df[['cell_name', 'age', 'location', 'Senescence_score']]\n",
    "\n",
    "# Export to CSV\n",
    "final_df.to_csv('senescence_scores.csv', index=False)\n",
    "\n",
    "print(\"Data has been written to senescence_scores.csv\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "14a830c3-c91f-4d79-b381-d0aea1c74371",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define the colors\n",
    "colors = ['#F08205', '#4C026E']\n",
    "\n",
    "# Create the seaborn box plot with specified colors and adjust box width and outlier size\n",
    "ax = sns.boxplot(data=final_df, x=\"location\", y='Senescence_score', hue=\"age\", palette=colors, width=0.5, fliersize=2)\n",
    "\n",
    "# Make outliers slightly transparent\n",
    "for artist in ax.artists:\n",
    "    artist.set_edgecolor('black')  # Set outlier edge color to black\n",
    "    for line in artist.get_lines():\n",
    "        line.set_alpha(0.1)  # Set outlier transparency to 0.5\n",
    "\n",
    "# Remove legends\n",
    "#plt.legend().remove()\n",
    "\n",
    "# Set figure size\n",
    "plt.gcf().set_size_inches(10, 5)\n",
    "\n",
    "# Customize y-axis display\n",
    "plt.yticks(fontsize=7)\n",
    "plt.gca().set_ylim([-0.65, 1.05])  # Set the y-axis limits\n",
    "plt.gca().set_yticks([-0.6,-0.4,-0.2, 0, 0.2, 0.4, 0.6, 0.8,1.0])  # Set y-axis ticks\n",
    "\n",
    "#Give angle to x labels\n",
    "plt.xticks(rotation=45, ha='right', fontsize=7)\n",
    "# Save the figure as PDF with high resolution\n",
    "#plt.savefig(\"box_plot.pdf\", dpi=300, bbox_inches='tight')\n",
    "\n",
    "# Show the plot (optional)\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3f483fc1-ee86-401d-b246-4b2f0bc3841f",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Save source data for figure "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b855cf94-9763-4832-bc86-fc93a4b88588",
   "metadata": {},
   "source": [
    "### Full size spatial plots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e753e8e9-2458-427d-8fc3-8c2ebfa61e27",
   "metadata": {},
   "outputs": [],
   "source": [
    "titles = ['Young', 'Geriatric']\n",
    "\n",
    "plot_spatial_data(adata_dict, obs=['max_pred_celltype'], titles=titles,  spot_size=15)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c27d3808-e05d-4323-b546-313534762e6e",
   "metadata": {},
   "outputs": [],
   "source": [
    "titles = ['Young', 'Geriatric']\n",
    "\n",
    "plot_spatial_data(adata_dict, obs=['injury_classification'], titles=titles,  spot_size=15)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6aadf109-9e8c-42ff-b788-2b7acff16686",
   "metadata": {},
   "source": [
    "### Zoomed-in plots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "840416f3-80e8-438a-831f-06cc1119843a",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "crop = {0: [1500, 2800, 2000, 700], 1: [1500, 2800, 2800, 1500]}\n",
    "titles = ['Young', 'Geriatric']\n",
    "\n",
    "plot_spatial_data(adata_dict, obs=['max_pred_celltype'], titles=titles, crop=crop, spot_size=15)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "903859df-0be9-40f7-a3ac-98ed128ed65a",
   "metadata": {},
   "outputs": [],
   "source": [
    "crop = {0: [1500, 2800, 2000, 700], 1: [1500, 2800, 2800, 1500]}\n",
    "titles = ['Young', 'Geriatric']\n",
    "plot_spatial_data(adata_dict, obs=['injury_classification'], titles=titles, crop=crop,spot_size=15)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c71efa18-7f9d-42e5-b692-833042b6f783",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "crop = {0: [1500, 2800, 2000, 700], 1: [1500, 2800, 2800, 1500]}\n",
    "titles = ['Young', 'Geriatric']\n",
    "plot_spatial_data(adata_dict, obs=['Senescence_score'], titles=titles, crop=crop, cmap=newcmp4, vmin=0.0000001, vmax=0.8, spot_size=15)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "437c600b-5611-49dd-aa32-630f62afb30c",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_dict[0].write_h5ad('/fs/cbsuvlaminck2/workdir/in68/Muscle/out/Sample11/Young_TA_ST_5dpi_pp.h5ad')\n",
    "adata_dict[1].write_h5ad('/fs/cbsuvlaminck2/workdir/in68/Muscle/out/Sample12/Geriatric_TA_ST_5dpi_pp.h5ad')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Environment (cell2loc_env)",
   "language": "python",
   "name": "cell2loc_env"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.17"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
